### PR TITLE
JAMES-2287 HashBlobId.Factory support encoding type

### PR DESCRIPTION
**Pull request by vttranlina**
- The base64 encoding has the `/` slash character that is unsupported in some S3 platform

### CODE REVIEW

**Comment by chibenwa** [comment]
To be fully serious please reproduce the invalid char on top of minio in a test. 

We would also benefit from running the all blob store suite against minio IMO.

**Comment by chibenwa** [comment]
(Good work but lets be sure we solved the customer issue by adding tests)

**Review by Arsnael — APPROVED** [review]
Looks good to me

**Review by chibenwa — COMMENTED** [review]
Just add a little test to ensure that now we can save 1000 different hashBlobId

File: server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3MinioTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+@Testcontainers
+public class S3MinioTest {
```
Comment (chibenwa) [inline server/blob/blob-s3/src/test/java/org/apache/james/blob/objectstorage/aws/S3MinioTest.java:48]: Any chance to extend BlobStoreDAOContract?

**Review by chibenwa — APPROVED** [review]

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/HashBlobId.java [hunk 2]
```diff
@@ -29,11 +29,22 @@
...
+    private static final String HASH_BLOB_ID_ENCODING_TYPE_PROPERTY = "james.blob.id.hash.encoding";
+    private static final BaseEncoding HASH_BLOB_ID_ENCODING_DEFAULT = BaseEncoding.base64Url();
 
     public static class Factory implements BlobId.Factory {
+        private final BaseEncoding baseEncoding;
+
+        public Factory() {
+            this.baseEncoding = Optional.ofNullable(System.getProperty(HASH_BLOB_ID_ENCODING_TYPE_PROPERTY))
+                    .map(Factory::baseEncodingFrom)
+                    .orElse(HASH_BLOB_ID_ENCODING_DEFAULT);
```
Comment (quantranhong1999) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/HashBlobId.java:45]: Should be 4 spaces?

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

File: server/blob/blob-api/src/main/java/org/apache/james/blob/api/HashBlobId.java (hunk 2 above)
Comment (chibenwa) [inline server/blob/blob-api/src/main/java/org/apache/james/blob/api/HashBlobId.java:45]:
```suggestion
            this.baseEncoding = Optional.ofNullable(System.getProperty(HASH_BLOB_ID_ENCODING_TYPE_PROPERTY))
                .map(Factory::baseEncodingFrom)
                .orElse(HASH_BLOB_ID_ENCODING_DEFAULT);
```

**Pull request merged**